File: apps/core-api/lib/lotta/tenants/feedback.ex
defmodule Lotta.Tenants.Feedback do
  @moduledoc """
  Ecto Schema for feedbacks
  """

  use Ecto.Schema

  import Ecto.Changeset

  @type id() :: binary()

  @type t() :: %__MODULE__{
          id: id(),
          user_id: id(),
          topic: String.t(),
          content: String.t(),
          is_new: boolean(),
          is_forwarded: boolean(),
          is_responded: boolean(),
          metadata: String.t(),
          inserted_at: DateTime.t(),
          updated_at: DateTime.t()
        }

  @primary_key {:id, :binary_id, autogenerate: true}

  @timestamps_opts [type: :utc_datetime]

  schema "feedbacks" do
    field :topic, :string
    field :content, :string
    field :is_new, :boolean
    field :is_forwarded, :boolean
    field :is_responded, :boolean
    field :metadata, :string

    belongs_to :user, Lotta.Accounts.User

    timestamps()
  end

  def create_changeset(%__MODULE__{} = feedback, user, attrs) do
    feedback
    |> cast(attrs, [:topic, :content, :metadata])
    |> put_assoc(:user, user)
    |> validate_required([:topic, :content])
  end

  def set_is_forwarded_changeset(%__MODULE__{} = feedback) do
    feedback
    |> cast(%{is_forwarded: true}, [:is_forwarded])
    |> validate_required([:is_forwarded])
    |> validate_change(:is_forwarded, fn _ ->
      [is_forwarded: "Feedback wurde bereits weitergeleitet."]
    end)
  end
end

File: packages/webapp/src/test/fixtures/Feedback.ts
import { FeedbackModel } from 'model';

export const newFeedback: FeedbackModel = {
  id: '6543-feed-back-1234',
  insertedAt: new Date().toISOString(),
  updatedAt: new Date().toISOString(),
  isForwarded: false,
  isResponded: false,
  isNew: true,
  topic: 'Feedback A',
  content: 'Feedback A content',
};

export const readFeedback: FeedbackModel = {
  id: '6973-feed-back-1432',
  insertedAt: new Date().toISOString(),
  updatedAt: new Date().toISOString(),
  isForwarded: false,
  isResponded: false,
  isNew: false,
  topic: 'Feedback B',
  content: 'Feedback B content',
};

export const respondedFeedback: FeedbackModel = {
  id: '1111-feed-back-1234',
  insertedAt: new Date().toISOString(),
  updatedAt: new Date().toISOString(),
  isForwarded: false,
  isResponded: true,
  isNew: false,
  topic: 'Feedback C',
  content: 'Feedback C content',
};

export const forwardedFeedback: FeedbackModel = {
  id: '9999-feed-back-1234',
  insertedAt: new Date().toISOString(),
  updatedAt: new Date().toISOString(),
  isForwarded: true,
  isResponded: false,
  isNew: false,
  topic: 'Feedback D',
  content: 'Feedback D content',
};

export const feedbacks = [
  newFeedback,
  readFeedback,
  respondedFeedback,
  forwardedFeedback,
];
